### PR TITLE
Fix DER reading from stdin for BIO_f_readbuffer

### LINKED ISSUE #14559: 3.0.0-alpha14-dev can't read DER from stdin
#14407 seems to have fixed the problem of reading input from stdin only if the input is PEM encoded. However, 3.0.0-alpha14-dev still cannot handle reading input from stdin if the input is DER encoded.

An (non-realistic) example of the problem is the following:
```
openssl x509 -in ./demos/bio/root.pem -outform DER | openssl x509 -inform DER -text -noout
```
This works with OpenSSL 1.1.1 and earlier, but with 3.0.0-alpha14-dev the result is:
```
Could not read certificate from <stdin>
809B26CAB07F0000:error:0680008E:asn1 encoding routines:asn1_d2i_read_bio:not enough data:crypto/asn1/a_d2i_fp.c:198:
Unable to load certificate
```
Similarly, the following works with OpenSSL 1.1.1 and earlier, but not with 3.0.0-alpha14-dev:
```
openssl x509 -in ./demos/bio/root.pem -pubkey -noout | openssl pkey -pubin -outform DER | openssl pkey -pubin -inform DER -text_pub
```
With OpenSSL 3.0.0-alpha14-dev the result is:
```
Could not read key of Public Key from <stdin>
809B0DAF9B7F0000:error:0680008E:asn1 encoding routines:asn1_d2i_read_bio:not enough data:crypto/asn1/a_d2i_fp.c:198:
```
The error is created by the third call to OpenSSL. The second call, which receives a PEM-encoded public key as input, works correctly with OpenSSL 3.0.0-alpha14-dev.

### PR DESCRIPTION
**Pull request by slontis**
Fixes #14559

The intitial implementation of the gets() function tried using the next bio's gets() function.
For a file BIO this returned incorrect data for binary data containing 0x00.
Just buffering all data during gets() did not work however since some
applications open and close the bio multiple times when dealing with pem
files containing multiple entries.. This does not work
when reading from stdin unless the data if buffered one byte at a time.

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
Manually verified that previous issues raised also still work... i.e.

````
> cat test/certs/leaf.pem | ./apps/openssl x509 -MD5 -fingerprint -noout -inform PEM
> cat test/recipes/20-test_dhparam_check_data/dhx_ffdhe2048.pem | ./apps/openssl dhparam -noout -text
````

File: test/recipes/61-test_bio_readbuffer.t [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+ok(run(app([ 'openssl', 'x509', '-inform', 'PEM', '-in', $pemfile,
+             '-outform', 'DER', '-out', $derfile])),
+   "Generate a DER certificate");
+
+ok(run(test(["bio_readbuffer_test", $derfile])),
+        "running bio_readbuffer_test $derfile");
+
+ok(run(test(["bio_readbuffer_test", $pemfile])),
+        "running bio_readbuffer_test $pemfile");
```
Comment (levitte) [inline test/recipes/61-test_bio_readbuffer.t:29]: Add a line ending, please

File: test/recipes/61-test_bio_readbuffer.t [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+my $derfile = 'readbuffer_leaf.der';
+
+plan tests => 3;
+
+ok(run(app([ 'openssl', 'x509', '-inform', 'PEM', '-in', $pemfile,
+             '-outform', 'DER', '-out', $derfile])),
+   "Generate a DER certificate");
+
+ok(run(test(["bio_readbuffer_test", $derfile])),
+        "running bio_readbuffer_test $derfile");
```
Comment (levitte) [inline test/recipes/61-test_bio_readbuffer.t:26]: Nit: this quote should be aligned with the outermost parenthesis.  The same goes for the next test case

**Review by levitte — CHANGES_REQUESTED** [review]

File: crypto/bio/bf_readbuff.c [hunk 3]
```diff
@@ -220,49 +220,69 @@ static int readbuffer_gets(BIO *b, char *buf, int size)
...
-                *buf = '\0';
-                return num > 0 ? num : num_chars;
+    /* If data is already buffered then use this first */
+    if (ctx->ibuf_len > 0) {
+        p = &(ctx->ibuf[ctx->ibuf_off]);
+        found_newline = 0;
+        for (num_chars = 0;
+             (num_chars < ctx->ibuf_len) && (num_chars < size);
+             num_chars++) {
+            *(buf++) = p[num_chars];
```
Comment (paulidale) [inline crypto/bio/bf_readbuff.c:238]: consider dropping the parenthesis: it's `*buf++`

File: crypto/bio/bf_readbuff.c [hunk 4]
```diff
@@ -220,49 +220,69 @@ static int readbuffer_gets(BIO *b, char *buf, int size)
...
+      */
+     p = &(ctx->ibuf[ctx->ibuf_off]);
+     for (i = 0; i < (1 + size); ++i) {
+         j = BIO_read(b->next_bio, p, 1);
+         if (j <= 0) {
+             BIO_copy_next_retry(b);
+             *buf = '\0';
+             return num > 0 ? num : j;
+         }
+         *(buf++) = *p;
```
Comment (paulidale) [inline crypto/bio/bf_readbuff.c:279]: Same here.

File: crypto/bio/bf_readbuff.c [hunk 5]
```diff
@@ -220,49 +220,69 @@ static int readbuffer_gets(BIO *b, char *buf, int size)
...
+     /*
+      * Read more data from the next bio using BIO_read_ex:
+      * Note we cannot use BIO_gets() here as it does not work on a
+      * binary stream that contains 0x00. (Since strlen() will stop at
+      * any 0x00 not at the last read '\n' in a FILE bio).
+      * Also note that some applications open and close the file bio
+      * multiple times and need to read the next available block when using
+      * stdin - so we need to READ one byte at a time!
+      */
+     p = &(ctx->ibuf[ctx->ibuf_off]);
```
Comment (paulidale) [inline crypto/bio/bf_readbuff.c:271]: The parenthesis here are option too, although `p = ctx->ibuf + ctx->ibuf_off;` is even more succinct.

**Comment by slontis** [comment]
looks like I have an off by 1 error here..

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]
I did a double take on that `1 +` but figured you knew what you were doing....

**Comment by slontis** [comment]
> I did a double take on that `1 +` but figured you knew what you were doing....

Obviously not :)

**Comment by t8m** [comment]
This is ready now

**Comment by slontis** [comment]
Thanks for reviewing. Merged to master.

**Pull request closed without merging**